### PR TITLE
Implement a church-notation-free Maybe type (as an Optional replacement)

### PR DESCRIPTION
**Pull request by ashfordneil**
Related Ticket:

# Description of Changes:
TL;DR => does what it says on the label. While it's really nice to be able to write left-to-right code with `opt.bind(...)`, I strongly believe that the pros (listed below) outweigh the cons here.

## Type guards
Look at the implementation of `lift5` vs the old implementation of `lift5`. We now have no need for `.getOrDie()` because typescript will automatically infer that if `myMaybe.tag === 'JUST'` then `myMaybe.value` has to be defined. This basically comes for free when we do ADTs the way [typescript recommends](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) instead of with church notation.

We also don't really need an explicit `isSome` or `isNone` function (I can add some if there's a demand), as the `.tag` is public and typescript will automatically call you out if you try something like `myMaybe.tag === 'JUUST'`.

## Variance ([background](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)))
In an ideal world, `Optional<Cat>` and `Optional<Animal>` should have the same type relationship as `Cat` and `Animal` (ie `Optional` should be covariant). Anywhere that an `Optional<Animal>` can be used, an `Optional<Cat>` should be usable instead. **However** because of Church notation, that breaks at the moment.

- The `myOpt.getOrDie()` function is _covariant_, as any use of `getOrDie()` that expects an `Optional<Animal>` will still work if you give it an `Optional<Cat>`. This is because in an `Optional<T>`, `T` is the return type of `getOrDie()`
- The `myOpt.is()` function is _contravariant_, any use of `is()` that expects an `Optional<Cat>` will work if you give it an `Optional<Animal>`. This is because in an `Optional<T>`, `T` is the argument type of `is()`
- Because both of these types are shipped with the same `Optional<T>`, the type as a whole becomes invariant.

Because the methods are no longer attached to the `Optional<T>`, this is no longer a problem at all. Compare the pair:

```typescript
const optional: Optional<Animal> = makeOpt<Cat>();
/*
Type 'Optional<Cat>' is not assignable to type 'Optional<Animal>'..
  Types of property 'is' are incompatible.
    Type '(input: Cat) => boolean' is not assignable to type '(input: Animal) => boolean'.
      Types of parameters 'input' and 'input' are incompatible.
        Type 'Animal' is not assignable to type 'Cat'.
          Type 'Dog' is not assignable to type 'Cat'
*/
```

vs this

```typescript
const maybe: Maybe<Animal> = makeMaybe<Cat>();
const maybe2: Maybe<Cat> = makeMaybe<Cat>();
Maybe.is(maybe, makeCat()); // Maybe.is<Animal>
Maybe.is(maybe2, makeCat()); // Maybe.is<Cat>
Maybe.is(maybe, makeAnimal()); // Maybe.is<Animal>
Maybe.is(maybe2, makeAnimal()); // Maybe.is<Animal>

/* No errors */
```

## Performance / Tree Shaking / Minifying
Because all the functions are standalone, we should see some _slight_ improvements to performance, some _pretty decent_ improvements to tree shaking, and some _alright_ improvements to minification. We also have the option (tempered by readability) to use regular if statements rather than things like `.fold()` or `.each()` if we so choose, while still retaining type safety.

## Debugging
Maybe instances can:
- Be logged, normally
- Be inspected in the debugger, normally

# Pre-checks:
* ~[ ] Changelog entry added~
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)
* ~[ ] License headers added on new files (if applicable)~

# Review:
* [ ] Milestone set
* [ ] Review comments resolved

GitHub issues (if applicable): N/A

### CODE REVIEW

File: modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+// Foldable / array style things
+export const exists = <T>(self: Maybe<T>, pred: (input: T) => boolean): boolean =>
+  self.tag === 'JUST' && pred(self.value);
+
+export const forAll = <T>(self: Maybe<T>, pred: (input: T) => boolean): boolean =>
+  self.tag === 'NOTHING' || pred(self.value);
+
+export const filter: {
+  <T>(self: Maybe<T>, pred: (input: T) => boolean): Maybe<T>;
+  <T, U extends T>(self: Maybe<T>, pred: (input: T) => input is U): Maybe<U>;
```
Comment (tiny-james) [inline modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts:171]: I thought typescript tried applying types top to bottom, in which case this would not gain the type filtering. Have you checked this applies the type filtering?

**Review by tiny-james — APPROVED** [review]
One other suggestion is to add `resultValue.toMaybe()` and `Result.fromMaybe(maybeValue)` to Result.

**Comment by spocke** [comment]
I want to see some concrete examples of using this in replacement of Optional. I think it would be very hard to read code with lots of chained calls or it would have to be broken up into variables and then the new code would be rewritten code would be bigger unless the minifier is smart and inlines the values. How important is support for subtyping do we use that a lot? Is there a concern that value is exposed, not sure we can easily hide it but it is a difference from Optional? It would be major work to switch everything to this it might be better to wait until we switch language there isn't much stopping use from using reasonml today for new plugins etc. The chaining would be a non issue there `filter(bind(map(opt, func), func2), func3)` would just be written as `opt->map(func)->bind(func2)->filter(func3)` with the fast pipe but compiled to the former format. With this maybe type it would either have to be manually written that way or needed to be split up into variables and multiple rows. I'm still not sold on this change I think the new problems with this approach is higher than the problems we have with the Optional type but if everyone else loves this I guess it just have to live with it. :)

File: modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  readonly tag: 'JUST';
+  readonly value: T;
+}
+
+export interface Nothing {
+  readonly tag: 'NOTHING';
+}
+
+// Constructors
+export const just = <T>(value: T): Just<T> => ({
```
Comment (ashfordneil) [inline modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts:13]: Worth mentioning: at present this can cause some issues with tsc inference. The problem revolves around "widening", explained below.

Typescript will automatically **widen** `"specific-string"` to `string` in it's type checking **in certain situations**. These situations include things like:
- when you use a `let` binding instead of const (`let thingo: "specific-string" = "specific-string"` makes it a bit hard to mutate)
- when you put in in an array with other strings
- when you have an expression that will conditionally evaluate to either `"specific-string"` or a different string
(note that in all of these situations you can explicitly ask typescript to not widen the type, if you choose)

Because tsc can automate the widening process at the times you'd want to widen things, they can safely default to the narrowest types possible, and trust that instead of something like this:
```typescript
type "other-string" not assignable to type "specific-string"
```
you won't get any errors (as the type widener has inferred that it's just a `string`)

Typescript will do similar widening for array literals (`[string, number] => (string | number)[]`) and for numeric literals (`3 => number`) where it sees fit. Unfortunately, we can't persuade typescript to do automatic type widening for custom types.

Consider the following code:
```typescript
let state = Maybe.just(3);
// do things
state = Maybe.bind(state, (inner) => ...)
```

This code won't compile, because `state` is of type `Just<number>` instead of type `Maybe<number>`, and therefore trying to assign a `Maybe<number>` to it (the return type of bind) will fail under the claim `type 'Nothing' is not assignable to type 'Just<number>'`. In an ideal world, Typescript would see that this was a `let` binding, and automatically assign the type to `Maybe<number>` instead of `Just<number>`. That's not an option, which leaves us with the following two choices:

## Choice 1: Manual type widening
By being explicit about the types, we can tell Typescript that things are widened manually, sidestepping the "how do we infer it" problem. This would involve the following types of changes:
- `let state = Maybe.just(3)` ==> `let state: Maybe<number> = Maybe.just(3)`
- Consider this hypothetical code:
```typescript
const toMaybe = <T>(opt: Optional<T>): Maybe<T> => opt.fold(Fun.constant(Maybe.Nothing), Maybe.just)
```
We would need to be change `opt.fold(...` to `opt.fold<Maybe<T>>(...` lest the compiler complain that the two arguments of `fold` return different types.
- `const x = [just(3), just(5), nothing]` ==> `const x: Array<Maybe<number>> = [just(3), just(5), nothing]`

## Choice 2: Initial type widening
Rather than having the `just` function returning the `Just` type, we could just make it return `Maybe`. This would avoid all of those problems, but it would also make things less specific / less technically correct. Hopefully, though, the type of inferences this change would break ("hey this `Maybe` was created with the `just` constructor, so it'll always have a value") don't come up in real code (if there's no conditionality in the code, or you're guaranteed to be using the same constructor any time that code is executed, why are you using "Maybe" at all?).

## Choice 3: See if upstream Typescript supports this, or make a request for them to support this
I don't think this is a great idea, but I thought I'd mention it in case someone else thinks we should do this. Choice 2 seems like a good solution for us (although it is technically suboptimal, it is practically usable), so I don't see the need for implementing this _for our usecase_.

File: modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts (hunk 1 above)
Comment (ashfordneil) [reply modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts]: Good catch: this is fixed

File: modules/robin/src/main/ts/ephox/robin/zone/ZoneWalker.ts [hunk 3]
```diff
@@ -95,16 +95,19 @@ const walk = <E, D> (
...
-    state.each((state) => visit(universe, stack, transform, viewport, state));
+  let state = Fun.pipe(Maybes.just<Traverse<E>>({ item: start, mode: Gather.advance }),
+    (m) => Maybes.filter(m, shouldContinue)
+  );
 
-    state = state
-      .bind((state) => getNextStep(universe, viewport, state))
-      .bind((traverse) => Gather.walk(universe, traverse.item, traverse.mode, Gather.walkers().right()))
-      .filter(shouldContinue);
+  while (state.tag === 'JUST') {
```
Comment (JamesToohey) [inline modules/robin/src/main/ts/ephox/robin/zone/ZoneWalker.ts:103]: I do think it would be nicer to have `Maybes.isJust()`

**Comment by ashfordneil** [comment]
Continued in #6477
